### PR TITLE
JP-2113: Update NIRCam WFSS transforms to v.4 from GRISMCONF files

### PR DESCRIPTION
**Pull request by tapastro**
Resolves [JP-2113](https://jira.stsci.edu/browse/JP-2113)

Closes #

This PR modifies existing transforms for NIRCAM WFSS data to enable use of the newest GRISMCONF files created by @NorPirzkal. This most recent set of files increased the complexity of the spectral trace fit, such that the analytic inverse of the fit no longer exists for first order spectra, requiring the use of interpolation.

Note that the code (in the most recent commit) no longer supports old specwcs reference files - backward compatibility could be added back in if needed.

**WIP**

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by hbushouse** [comment]
Looks like lots of unit tests need updating (lots of failures in all CI tests)

**Comment by hbushouse** [comment]
I see new specwcs ref files delivered to CRDS-TEST on 11-7-2022, but no new photom ref files yet with the updated sensitivity curves for the grisms.

**Comment by bhilbert4** [comment]
@hbushouse photom files exist, but I don't think they've been scientifically validated yet. I can try to do that soon, although I'm not the expert on this.

Also, @NorPirzkal told me the other day that there is a v5 of the grismconf files available now. I'm assuming that doesn't really matter for testing the pipeline updates, but I thought I'd mention that we'll want to deliver updates to the reference files either before or immediately after everything is pushed to OPS.

**Comment by hbushouse** [comment]
Thanks @bhilbert4. If the v5 ref files are at least the same format and use same arrangement of polynomials, etc. as v4, then it should be straightforward to upgrade to v5 after we get v4 working.

**Comment by bhilbert4** [comment]
@hbushouse I'm working on the photom verification. Comparing the extracted spectrum for P330-E to a reference spectrum, I'm seeing the extracted spectrum with values that are larger by a factor of ~10. (I'm fairly sure I'm looking at the correct source in the x1d file.) Nor says the factor of 10 could come from not dividing the extracted spectrum by the width in Angstroms of the pixel. The sensitivity is per Angstrom and not per pixel. Does that agree with what the pipeline is doing?

**Comment by hbushouse** [comment]
> @hbushouse I'm working on the photom verification. Comparing the extracted spectrum for P330-E to a reference spectrum, I'm seeing the extracted spectrum with values that are larger by a factor of ~10. (I'm fairly sure I'm looking at the correct source in the x1d file.) Nor says the factor of 10 could come from not dividing the extracted spectrum by the width in Angstroms of the pixel. The sensitivity is per Angstrom and not per pixel. Does that agree with what the pipeline is doing?

The photom step applies the flux (surface brightness) conversion to 2D data by simply interpolating the wavelength-dependent relative response (sensitivity) function given in the photom ref file table to the (central) wavelength of each pixel in the 2D image, which for WFSS data is all the individual 2D source cutouts. So there's no accounting for the wavelength range covered by each pixel. Hence if the original sensitivity curve that's been derived is in units of per Angstrom, then I guess it would need to be converted (scaled by the per-pixel wavelength width) for storage in the photom ref file.

Works exactly the same way for NIRISS WFSS, so should be able to do whatever they did for their WFSS photom ref file data.

**Comment by hbushouse** [comment]
Started a regtest run at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/676/ that uses this PR branch, stdatamodels/master (which has the transforms updates), and crds-test context 1072, which has the new NIRCam specwcs ref files.

**Comment by hbushouse** [comment]
@tapastro Have a look through https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/676/testReport/  There are a lot of differences in what should be unrelated products (e.g. miri imaging), but that may be due to different ref files (crds-test 1072 is now a ways behind the latest/greatest), and there are definitely a few that show different ref files recorded in the header keywords, so we just ignore those. Some of the other differences that don't explicitly make mention of different ref files may be due to code changes, either in jwst or stcal, that aren't incorporated into this branch. It looks like at least some of the NIRCam WFSS tests ran successfully, but showed differences (as expected). A few things had actual errors. Not sure what's causing those.

**Comment by hbushouse** [comment]
Large number of differences in regtest run 676 have been traced to the fact that crds-test context 1072 has the snowball/shower correction turned ON in the jump step. Starting another run using the default/latest context to at least look for unintended consequences in all non-nircam-wfss tests. The nircam-wfss tests will fail due to the use of old specwcs ref files. https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/677

**Comment by hbushouse** [comment]
The 677 results look like all of the failures and/or errors were involving grism/wfss tests (both unit tests and regtests), and no other non-grism/non-wfss failures. So that suggests this PR is "safe" for non-WFSS code.

**Comment by hbushouse** [comment]
I ran https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/700/ using the latest stdatamodels and this PR branch, also using CRDS-TEST context 1078, which includes the new specwcs ref files. The NIRCam WFSS tests appear to have used the new specwcs ref files and resulted in many differences (as expected). But still also getting errors in some of the other NIRCam grism-related tests, such as the wfss_contam test and tsgrism (time-series grism) test.  Other failures are due to other recent PR mergers (e.g. update to NIRSpec pathloss correction).

**Comment by tapastro** [comment]
Regtests using CRDS-TEST references: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/712/#showFailuresLink

The two remaining failure groups I'm looking at after fixing the assign_wcs unit tests are the wfss_contam failures and the extract_2d failures. 

@nden I think the extract_2d failures are a result of the new references, though as a unit test I was worried that they shouldn't depend on the references. I might be off in assuming that, though.

@hbushouse I'm going to start taking a look at the wfss_contam failure - it looks like a failure case that may be expected for some cases, but the error trace may not be informative enough on its own here.

File: requirements-max.txt [hunk 1]
```diff
@@ -1,4 +1,3 @@
 # this requirements file applies upper pins on critical upstream dependencies with incompatible functionality / refactoring that has not yet been implemented in this repository
 
 stcal<1.4.0
-stdatamodels<1.4.1
```
Comment (zacharyburnett) [inline requirements-max.txt:4]: `stdatamodels 1.5.0` is now released, so this could be updated to `stdatamodels<1.6.0`

Comment (tapastro) [reply requirements-max.txt:4]: ✅

**Review by hbushouse — APPROVED** [review]
LGTM

**Review by nden — APPROVED** [review]

**Comment by zacharyburnett** [comment]
Lots of API errors in `test-oldestdeps-xdist`:
```
E           TypeError: __init__() got an unexpected keyword argument 'inv_lmodels'
```
When was `__init__` changed?

**Comment by tapastro** [comment]
> Lots of API errors in `test-oldestdeps-xdist`: https://github.com/spacetelescope/jwst/actions/runs/4995603889/jobs/8947685758?pr=7018#step:10:536
> 
> ```
> E           TypeError: __init__() got an unexpected keyword argument 'inv_lmodels'
> ```
> 
> When was `__init__` changed? We can update the lower pins to reflect that
> 
> EDIT: it looks like this is a new parameter of `stdatamodels` `1.5.0`; we should pin `stdatamodels>=1.5.0` in `setup.cfg`

Ah, my email didn't show your edit and I started worrying! But yes, I think we need to ensure this `jwst` requires 1.5.0. I'll get that committed.

**Review by zacharyburnett — APPROVED** [review]

**Pull request merged**